soroban-env-common/src/compare.rs

```diff
@@ -116,25 +116,33 @@ impl<T, C: Compare<T>> Compare<Rc<T>> for C {
 // better). But we can list out any concrete Ord instances we want to support
 // here.
 
-impl<E: Env> Compare<Object> for E {
-    type Error = E::Error;
-
-    fn compare(&self, a: &Object, b: &Object) -> Result<Ordering, Self::Error> {
-        let v = self.obj_cmp(a.to_raw(), b.to_raw())?;
-        if v == 0 {
-            Ok(Ordering::Equal)
-        } else if v < 0 {
-            Ok(Ordering::Less)
-        } else {
-            Ok(Ordering::Greater)
-        }
-    }
+macro_rules! delegate_compare_to_wrapper {
+    ($T:ident,$A:ident,$B:ident,$SELF:ident) => {{
+        let a = unsafe { crate::$T::unchecked_from_val(*$A) };
+        let b = unsafe { crate::$T::unchecked_from_val(*$B) };
+        $SELF.compare(&a, &b)
+    }};
 }
 
 impl<E: Env> Compare<RawVal> for E {
     type Error = E::Error;
 
     fn compare(&self, a: &RawVal, b: &RawVal) -> Result<Ordering, Self::Error> {
+        if a.get_payload() == b.get_payload() {
+            // Fast-path exactly-equal values.
+            return Ok(Ordering::Equal);
+        }
+        if a.is_object() || b.is_object() {
```

Comment:
Should this be && instead of ||?

Reply:
Hmm maybe not? Looks like `obj_cmp` was updated to handle non-Objects.

Reply:
`||` is correct here since `obj_cmp` handles all the cases where either one operand is an object.